apps/backend/src/lib/external-db-sync.ts

```diff
@@ -0,0 +1,203 @@
+import { Tenancy } from "@/lib/tenancies";
+import { getPrismaClientForTenancy, PrismaClientTransaction } from "@/prisma-client";
+import type { CompleteConfig } from "@stackframe/stack-shared/dist/config/schema";
+import { Client } from 'pg';
+
+export function getExternalDatabases(config: CompleteConfig) {
+  return config.dbSync.externalDatabases;
+}
+
+async function pushRowsToExternalDb(
+  externalClient: Client,
+  tableName: string,
+  newRows: any[],
+  upsertQuery?: string,
+) {
+  if (!upsertQuery) {
+    throw new Error(
+      `Cannot sync table "${tableName}": No upsertQuery configured.`
+    );
+  }
+
+  if (newRows.length === 0) return;
+  const placeholderMatches = upsertQuery.match(/\$\d+/g) ?? [];
+  const expectedParamCount =
+    placeholderMatches.length === 0
+      ? 0
+      : Math.max(...placeholderMatches.map((m) => Number(m.slice(1))));
+
+  if (expectedParamCount === 0) {
+    throw new Error(
+      `upsertQuery for table "${tableName}" contains no positional parameters ($1, $2, ...).` +
+        ` Your mapping must use parameterized SQL.`
+    );
+  }
+  const sampleRow = newRows[0];
+  const { tenancyId: _ignore, ...restSample } = sampleRow;
+  const orderedKeys = Object.keys(restSample);
+
+  if (orderedKeys.length !== expectedParamCount) {
+    throw new Error(
+      ` Column count mismatch for table "${tableName}".\n` +
+        `→ upsertQuery expects ${expectedParamCount} parameters.\n` +
+        `→ internalDbFetchQuery returned ${orderedKeys.length} columns (excluding tenancyId).\n` +
+        `Fix your SELECT column order or your SQL parameter order.`
+    );
+  }
+
+  for (const row of newRows) {
+    const { tenancyId, ...rest } = row;
+    const rowKeys = Object.keys(rest);
+
+    const validShape =
+      rowKeys.length === orderedKeys.length &&
+      rowKeys.every((k, i) => k === orderedKeys[i]);
+
+    if (!validShape) {
+      throw new Error(
+        `  Row shape mismatch for table "${tableName}".\n` +
+          `Expected column order: [${orderedKeys.join(", ")}]\n` +
+          `Received column order: [${rowKeys.join(", ")}]\n` +
+          `Your SELECT must be explicit, ordered, and NEVER use SELECT *.\n` +
+          `Fix the SELECT in internalDbFetchQuery immediately.`
+      );
+    }
+  }
+  for (const row of newRows) {
+    const { tenancyId, ...rest } = row;
+    await externalClient.query(upsertQuery, Object.values(rest));
+  }
```

Comment:
should we check that the tenancyId is always the same? i think right now we don't do anything with it, right?